File: internal/plugins/jest/rules/expect_expect/expect_expect_test.go
package expect_expect_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/expect_expect"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestExpectExpectRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&expect_expect.ExpectExpectRule,
		[]rule_tester.ValidTestCase{
			{Code: `it.todo("will test something eventually")`},
			{Code: `test.todo("will test something eventually")`},
			{Code: `['x']();`},
			{Code: `it("should pass", () => expect(true).toBeDefined())`},
			{Code: `test("should pass", () => expect(true).toBeDefined())`},
			{Code: `it("should pass", () => somePromise().then(() => expect(true).toBeDefined()))`},
			{Code: `it("should pass", myTest); function myTest() { expect(true).toBeDefined() }`},
			{Code: `function myTest() { expect(true).toBeDefined() } it("should pass", myTest);`},
			{
				Code: `
        function sharedAssertion() {
          expect(true).toBeDefined();
        }

        it("first passes", sharedAssertion);
        test("second passes", sharedAssertion);
      `,
			},
			{
				Code: `
        test('should pass', () => {
          expect(true).toBeDefined();
          foo(true).toBe(true);
        });
      `,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"expect", "foo"}}},
			},
			{
				Code:    `it("should return undefined",() => expectSaga(mySaga).returns());`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"expectSaga"}}},
			},
			{
				Code:    `test('verifies expect method call', () => expect$(123));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{`expect\$`}}},
			},
			{
				Code:    `test('verifies expect method call', () => new Foo().expect(123));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"Foo.expect"}}},
			},
			{
				Code: `
        test('verifies deep expect method call', () => {
          tester.foo().expect(123);
        });
      `,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"tester.foo.expect"}}},
			},
			{
				Code: `
        test('verifies chained expect method call', () => {
          tester
            .foo()
            .bar()
            .expect(456);
        });
      `,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"tester.foo.bar.expect"}}},
			},
			{
				Code: `
        test("verifies the function call", () => {
          td.verify(someFunctionCall())
        })
      `,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"td.verify"}}},
			},
			{
				Code: `it("should pass", () => expect(true).toBeDefined())`,
				Options: []interface{}{
					map[string]interface{}{
						"assertFunctionNames":          nil,
						"additionalTestBlockFunctions": nil,
					},
				},
			},
			{
				Code: `
        theoretically('the number {input} is correctly translated to string', theories, theory => {
          const output = NumberToLongString(theory.input);
          expect(output).toBe(theory.expected);
        })
      `,
				Options: []interface{}{map[string]interface{}{"additionalTestBlockFunctions": []interface{}{"theoretically"}}},
			},
			{
				Code:    `test('should pass *', () => expect404ToBeLoaded());`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"expect*"}}},
			},
			{
				Code:    `test('should pass *', () => expect.toHaveStatus404());`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"expect.**"}}},
			},
			{
				Code:    `test('should pass', () => tester.foo().expect(123));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"tester.*.expect"}}},
			},
			{
				Code:    `test('should pass **', () => tester.foo().expect(123));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"**"}}},
			},
			{
				Code:    `test('should pass *', () => tester.foo().expect(123));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"*"}}},
			},
			{
				Code:    `test('should pass', () => tester.foo().expect(123));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"tester.**"}}},
			},
			{
				Code:    `test('should pass', () => tester.foo().expect(123));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"tester.*"}}},
			},
			{
				Code:    `test('should pass', () => tester.foo().bar().expectIt(456));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"tester.**.expect*"}}},
			},
			{
				Code:    `test('should pass', () => request.get().foo().expect(456));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"request.**.expect"}}},
			},
			{
				Code:    `test('should pass', () => request.get().foo().expect(456));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"request.**.e*e*t"}}},
			},
			{
				Code: `test('should still lint with malformed pattern config', () => {
          expect(true).toBeDefined();
        });`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"(", "expect"}}},
			},
			{
				Code: `
        import { test } from '@jest/globals';

        test('should pass', () => {
          expect(true).toBeDefined();
          foo(true).toBe(true);
        });
      `,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"expect", "foo"}}},
			},
			{
				Code: `
        import { test as checkThat } from '@jest/globals';

        checkThat('this passes', () => {
          expect(true).toBeDefined();
          foo(true).toBe(true);
        });
      `,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"expect", "foo"}}},
			},
			{
				Code: `
        const { test } = require('@jest/globals');

        test('verifies chained expect method call', () => {
          tester
            .foo()
            .bar()
            .expect(456);
        });
      `,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"tester.foo.bar.expect"}}},
			},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: `it("should fail", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 3},
				},
			},
			{
				Code: `it("should fail", myTest); function myTest() {}`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 3},
				},
			},
			{
				Code: `test("should fail", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 5},
				},
			},
			{
				Code: `test.skip("should fail", () => {});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 10},
				},
			},
			{
				Code:    `afterEach(() => {});`,
				Options: []interface{}{map[string]interface{}{"additionalTestBlockFunctions": []interface{}{"afterEach"}}},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 10},
				},
			},
			{
				Code: `theoretically('the number {input} is correctly translated to string', theories, theory => {
          const output = NumberToLongString(theory.input);
        })`,
				Options: []interface{}{map[string]interface{}{"additionalTestBlockFunctions": []interface{}{"theoretically"}}},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 14},
				},
			},
			{
				Code: `it("should fail", () => { somePromise.then(() => {}); });`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 3},
				},
			},
			{
				Code:    `test("should fail", () => { foo(true).toBe(true); })`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"expect"}}},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 5},
				},
			},
			{
				Code:    `it("should also fail",() => expectSaga(mySaga).returns());`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"expect"}}},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 3},
				},
			},
			{
				Code:    `test('should fail', () => request.get().foo().expect(456));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"request.*.expect"}}},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 5},
				},
			},
			{
				Code:    `test('should fail', () => request.get().foo().bar().expect(456));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"request.foo**.expect"}}},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 5},
				},
			},
			{
				Code:    `test('should fail', () => tester.request(123));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"request.*"}}},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 5},
				},
			},
			{
				Code:    `test('should fail', () => request(123));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"request.*"}}},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 5},
				},
			},
			{
				Code:    `test('should fail', () => request(123));`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"request.**"}}},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 5},
				},
			},
			{
				Code:    `test('should fail without crashing on malformed pattern config', () => {});`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"("}}},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 5},
				},
			},
			{
				Code: `import { test as checkThat } from '@jest/globals';

        checkThat('this passes', () => {
          // ...
        });`,
				Options: []interface{}{map[string]interface{}{"assertFunctionNames": []interface{}{"expect", "foo"}}},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 3, Column: 9, EndLine: 3, EndColumn: 18},
				},
			},
			{
				Code: `import { test as checkThat } from '@jest/globals';

        checkThat.skip('this passes', () => {
          // ...
        });`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 3, Column: 9, EndLine: 3, EndColumn: 23},
				},
			},
			{
				Code: `it('outer', helper);
        function helper() {}

        describe('nested', () => {
          it('inner', helper);
          function helper() {
            expect(true).toBeDefined();
          }
        });`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noAssertions", Line: 1, Column: 1, EndLine: 1, EndColumn: 3},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/expect-expect.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('expect-expect', {} as never, {
  valid: [
    { code: 'it.todo("will test something eventually")' },
    { code: 'test.todo("will test something eventually")' },
    { code: "['x']();" },
    { code: 'it("should pass", () => expect(true).toBeDefined())' },
    { code: 'test("should pass", () => expect(true).toBeDefined())' },
    {
      code: 'it("should pass", () => somePromise().then(() => expect(true).toBeDefined()))',
    },
    {
      code: 'it("should pass", myTest); function myTest() { expect(true).toBeDefined() }',
    },
    {
      code: `
        test('should pass', () => {
          expect(true).toBeDefined();
          foo(true).toBe(true);
        });
      `,
      options: [{ assertFunctionNames: ['expect', 'foo'] }],
    },
    {
      code: 'it("should return undefined",() => expectSaga(mySaga).returns());',
      options: [{ assertFunctionNames: ['expectSaga'] }],
    },
    {
      code: "test('verifies expect method call', () => expect$(123));",
      options: [{ assertFunctionNames: ['expect\\$'] }],
    },
    {
      code: "test('verifies expect method call', () => new Foo().expect(123));",
      options: [{ assertFunctionNames: ['Foo.expect'] }],
    },
    {
      code: `
        test('verifies deep expect method call', () => {
          tester.foo().expect(123);
        });
      `,
      options: [{ assertFunctionNames: ['tester.foo.expect'] }],
    },
    {
      code: `
        test('verifies chained expect method call', () => {
          tester
            .foo()
            .bar()
            .expect(456);
        });
      `,
      options: [{ assertFunctionNames: ['tester.foo.bar.expect'] }],
    },
    {
      code: `
        test("verifies the function call", () => {
          td.verify(someFunctionCall())
        })
      `,
      options: [{ assertFunctionNames: ['td.verify'] }],
    },
    {
      code: 'it("should pass", () => expect(true).toBeDefined())',
      options: [
        {
          assertFunctionNames: undefined,
          additionalTestBlockFunctions: undefined,
        },
      ],
    },
    {
      code: `
        theoretically('the number {input} is correctly translated to string', theories, theory => {
          const output = NumberToLongString(theory.input);
          expect(output).toBe(theory.expected);
        })
      `,
      options: [{ additionalTestBlockFunctions: ['theoretically'] }],
    },

    // wildcards
    {
      code: "test('should pass *', () => expect404ToBeLoaded());",
      options: [{ assertFunctionNames: ['expect*'] }],
    },
    {
      code: "test('should pass *', () => expect.toHaveStatus404());",
      options: [{ assertFunctionNames: ['expect.**'] }],
    },
    {
      code: "test('should pass', () => tester.foo().expect(123));",
      options: [{ assertFunctionNames: ['tester.*.expect'] }],
    },
    {
      code: "test('should pass **', () => tester.foo().expect(123));",
      options: [{ assertFunctionNames: ['**'] }],
    },
    {
      code: "test('should pass *', () => tester.foo().expect(123));",
      options: [{ assertFunctionNames: ['*'] }],
    },
    {
      code: "test('should pass', () => tester.foo().expect(123));",
      options: [{ assertFunctionNames: ['tester.**'] }],
    },
    {
      code: "test('should pass', () => tester.foo().expect(123));",
      options: [{ assertFunctionNames: ['tester.*'] }],
    },
    {
      code: "test('should pass', () => tester.foo().bar().expectIt(456));",
      options: [{ assertFunctionNames: ['tester.**.expect*'] }],
    },
    {
      code: "test('should pass', () => request.get().foo().expect(456));",
      options: [{ assertFunctionNames: ['request.**.expect'] }],
    },
    {
      code: "test('should pass', () => request.get().foo().expect(456));",
      options: [{ assertFunctionNames: ['request.**.e*e*t'] }],
    },

    // alias
    {
      code: `
        import { test } from '@jest/globals';

        test('should pass', () => {
          expect(true).toBeDefined();
          foo(true).toBe(true);
        });
      `,
      options: [{ assertFunctionNames: ['expect', 'foo'] }],
    },
    {
      code: `
        import { test as checkThat } from '@jest/globals';

        checkThat('this passes', () => {
          expect(true).toBeDefined();
          foo(true).toBe(true);
        });
      `,
      options: [{ assertFunctionNames: ['expect', 'foo'] }],
    },
    {
      code: `
        const { test } = require('@jest/globals');

        test('verifies chained expect method call', () => {
          tester
            .foo()
            .bar()
            .expect(456);
        });
      `,
      options: [{ assertFunctionNames: ['tester.foo.bar.expect'] }],
    },
  ],
  invalid: [
    {
      code: 'it("should fail", () => {});',
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 3,
        },
      ],
    },
    {
      code: 'it("should fail", myTest); function myTest() {}',
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 3,
        },
      ],
    },
    {
      code: 'test("should fail", () => {});',
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 5,
        },
      ],
    },
    {
      code: 'test.skip("should fail", () => {});',
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 10,
        },
      ],
    },
    {
      code: 'afterEach(() => {});',
      options: [{ additionalTestBlockFunctions: ['afterEach'] }],
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 10,
        },
      ],
    },
    {
      code: `
        theoretically('the number {input} is correctly translated to string', theories, theory => {
          const output = NumberToLongString(theory.input);
        })
      `,
      options: [{ additionalTestBlockFunctions: ['theoretically'] }],
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 14,
        },
      ],
    },
    {
      code: 'it("should fail", () => { somePromise.then(() => {}); });',
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 3,
        },
      ],
    },
    {
      code: 'test("should fail", () => { foo(true).toBe(true); })',
      options: [{ assertFunctionNames: ['expect'] }],
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 5,
        },
      ],
    },
    {
      code: 'it("should also fail",() => expectSaga(mySaga).returns());',
      options: [{ assertFunctionNames: ['expect'] }],
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 3,
        },
      ],
    },

    // wildcards
    {
      code: "test('should fail', () => request.get().foo().expect(456));",
      options: [{ assertFunctionNames: ['request.*.expect'] }],
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 5,
        },
      ],
    },
    {
      code: "test('should fail', () => request.get().foo().bar().expect(456));",
      options: [{ assertFunctionNames: ['request.foo**.expect'] }],
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 5,
        },
      ],
    },
    {
      code: "test('should fail', () => tester.request(123));",
      options: [{ assertFunctionNames: ['request.*'] }],
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 5,
        },
      ],
    },
    {
      code: "test('should fail', () => request(123));",
      options: [{ assertFunctionNames: ['request.*'] }],
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 5,
        },
      ],
    },
    {
      code: "test('should fail', () => request(123));",
      options: [{ assertFunctionNames: ['request.**'] }],
      errors: [
        {
          messageId: 'noAssertions',
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 5,
        },
      ],
    },

    // aliases
    {
      code: `
        import { test as checkThat } from '@jest/globals';

        checkThat('this passes', () => {
          // ...
        });
      `,
      options: [{ assertFunctionNames: ['expect', 'foo'] }],
      errors: [
        {
          messageId: 'noAssertions',
          line: 3,
          column: 1,
          endLine: 3,
          endColumn: 10,
        },
      ],
    },
    {
      code: `
        import { test as checkThat } from '@jest/globals';

        checkThat.skip('this passes', () => {
          // ...
        });
      `,
      errors: [
        {
          messageId: 'noAssertions',
          line: 3,
          column: 1,
          endLine: 3,
          endColumn: 15,
        },
      ],
    },
  ],
});
